netket/nn/symmetric_linear.py

```diff
@@ -101,62 +101,446 @@ def full_bias(self, bias):
+    """Mask that zeros out elements of the filter. Should be of shape [inp.shape[-1]]"""
+    dtype: DType = jnp.float64
+    """The dtype of the weights."""
+    precision: Any = None
+    """numerical precision of the computation see `jax.lax.Precision`for details."""
+
+    kernel_init: Optional[Callable[[PRNGKeyT, Shape, DType], Array]] = None
+    """Initializer for the Dense layer matrix."""
+    bias_init: Optional[Callable[[PRNGKeyT, Shape, DType], Array]] = None
+    """Initializer for the bias."""
+
+    def setup(self):
+
+        self.n_cells = np.product(np.asarray(self.shape))
+        self.n_symm = len(np.asarray(self.product_table)) // self.n_cells
+
+        self.stdev = 1.0 / np.sqrt(self.in_features * self.n_cells * self.n_symm)
+
+        self.mapping = (
+            np.asarray(self.product_table)[: self.n_symm]
+            .reshape(self.n_symm, self.n_cells, self.n_symm)
+            .transpose(0, 2, 1)
+            .reshape(self.n_symm, self.n_symm, *self.shape)
+        )
+
+    def make_kernel(self, kernel):
+        kernel = kernel[..., self.mapping]
+
+        return kernel
+
+    @compact
+    def __call__(self, x: Array) -> Array:
+        """Applies the equivariant transform to the inputs along the last two
+        dimensions (-2: features, -1: group elements)
+        """
+
+        dtype = jnp.promote_types(x.dtype, self.dtype)
+        x = jnp.asarray(x, dtype)
+
+        x = x.reshape(*x.shape[:-1], self.n_cells, self.n_symm).transpose(0, 1, 3, 2)
+        x = x.reshape(*x.shape[:-1], *self.shape)
+
+        if self.kernel_init:
+            kernel = self.param(
+                "kernel",
+                self.kernel_init,
+                (
+                    self.out_features,
+                    self.in_features,
+                    self.n_symm * self.n_cells,
+                ),
+                self.dtype,
+            )
+        else:
+            kernel = self.param(
+                "kernel",
+                normal(self.stdev),
+                (
+                    self.out_features,
+                    self.in_features,
+                    self.n_symm * self.n_cells,
+                ),
+                self.dtype,
+            )
+
+        kernel = jnp.asarray(kernel, dtype)
+
+        if self.mask:
+            kernel = kernel * jnp.expand_dims(self.mask, (0, 1))
+
+        kernel = self.make_kernel(kernel)
+
+        x = jnp.fft.fftn(x, s=self.shape).reshape(*x.shape[:3], self.n_cells)
+
+        kernel = jnp.fft.fftn(kernel, s=self.shape).reshape(
+            *kernel.shape[:4], self.n_cells
+        )
+
+        x = lax.dot_general(
+            x, kernel, (((1, 2), (1, 2)), ((3,), (4,))), precision=self.precision
+        )
+        x = x.transpose(1, 2, 3, 0)
+        x = x.reshape(*x.shape[:3], *self.shape)
+
+        x = jnp.fft.ifftn(x, s=self.shape).reshape(*x.shape[:3], self.n_cells)
+        x = x.transpose(0, 1, 3, 2).reshape(*x.shape[:2], -1)
```

Comment:
This `x` is always complex, even if both the input and the kernels are real. You should cast it to the appropriate dtype as it is done in `DenseEquivariantIrrep`

Reply:
@attila-i-szabo  don't we want to use self.dtype not dtype here? Or is there something I"m missing.

```
if jnp.iscomplexobj(dtype):
    return jnp.asarray(inputs, dtype=dtype)
else:
    return jnp.asarray(inputs.real, dtype=dtype)
```

Reply:
I added this thing to fix complex input real self.dtype.

Reply:
I think this was Atilla's suggestion to make sure the output is the same as the dtype of the model (sometimes we want a real valued function but have to use an FFT or something). We still have the code below that makes sure that all of the arrays are the same dtype to avoid those errors

```
dtype = jnp.promote_types(x.dtype, self.dtype)
x = jnp.asarray(x, dtype)
kernel = jnp.asarray(kernel, dtype)
bias = jnp.asarray(kernel, dtype)
```

Reply:
>  don't we want to use self.dtype not dtype here?

No. The output is complex if the input _or_ the kernel is complex, so using the promoted `dtype` is appropriate

Reply:
Okay I understand. Anyway that function was returning some strange errors so I used

```
jnp.can_cast(x, dtype):
    return x
else: 
    return x.real 
```
Which if x is complex and dtype is complex this just returns x and if x is complex but dtype is real it returns x.real (which is automatically a float)